### ISSUE TITLE
stream data while copying deduped metadata between different ARBDs

### ISSUE THREAD

**Issue by GlenDC** [zero-os/0-Disk#353]
Currently when copying deduped metadata between different ARBDs (e.g. when running `zeroctl copy vdisk`, where the given source vdisk is deduped). We are fetching all the metadata (the LBA hashmap) to then store it on the other server, however this is quite memory intensive, and might break on low-memory devices.

For a `20 GiB` disk that would mean we are fetching `160 MiB` at once (`LBASize = diskSize / blockSize * hashSize`, where the hash size is `32 B`). This is by the way done currently using the `HGETALL` command. Obviously we also don't want to move to a world where for a 20 GiB we would fetch each LBA `sector` separately, and up with with like `40960` roundtrips for such a disk.

## How to fix it

Ideally we would simply use [the redis `MIGRATE` command](https://redis.io/commands/migrate) for this, so that we can simply stream the data between the 2 ARDBs, however we do not use redis servers directly AFAIK, or at least not always, and instead rely on [yinqiwen/ARDB](https://github.com/yinqiwen/ardb), which means we would instead need to use [the `MIGRATEDB` command](https://github.com/yinqiwen/ardb/wiki/ARDB-Commands#migratedb---use-to-migrate-data-in-current-db-to-other-server-only-works-with-remote-ardb-instance) it exposes. I see 2 potential issues with this. First of all, I'm not sure if this alternative command is as complete as redis's one, and second of all we would somehow need to detect if we are running ardb, redis, or any other server, so we can use only what is supported I guess. And of course the question is what if we want to migrate between an ardb and redis server. A lot of questions, and something that definitely have to be figured out.

Another (less ideal) approach could be to somehow stream the data from the source to the target server, going via the zeroctl user's machine. The simplest approach if we go for this direction would be to simply read static sized packages (e.g. `32 MiB`) , and write those packages to the target server as we receive them. This would avoid having to fetch all data first before writing, but it would still not be a true stream, as that is not something that github.com/garyburd/redigo/redis supports. Another approach in this direction could be that we do still just issue a command to fetch all data at once (e.g. `HGETALL`), but instead of waiting to receive it all, we already start streaming the data to the target server. In the end we are already using TCP which is a streaming protocol, so it should be possible. The redis protocol is also simpel enough that we could write a small module to allow for this.

TLDR: make sure we don't have to fetch the entire LBA at once into memory in order to copy it between different ARDB servers (e.g. by issuing the `zeroctl copy` command).

**Comment by GlenDC** [comment]
In order to never have too much data at once copied over, we'll use [`HSCAN`](https://redis.io/commands/hscan), it will allow us to iterate through the entire hash map (as we do now using [`HGETALL`](https://redis.io/commands/hgetall), but with the difference that we can limit the amount of data we'll fetch at once. This will make the process a bit slower, but it will make the process scalable to as many blocks as wanted, without getting huge memory-spikes during the copy process.
